### PR TITLE
demoit: init at unstable-2019-03-29

### PR DESCRIPTION
**Pull request by freezeboy**
###### Motivation for this change

Add `demoit` application from dgageot
It lets include a shell in web pages (used in html5 slides (bespoke, reveal.js, ...))

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/servers/demoit/default.nix [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{ stdenv
+, buildGoPackage
+, fetchFromGitHub
+}:
+
+buildGoPackage rec {
+  name = "demoit";
```
Comment (symphorien) [inline pkgs/servers/demoit/default.nix]: The name should include a version. If upstream does not release versions at all, then use the date in YYYYMMDD
format.

Comment (symphorien) [reply pkgs/servers/demoit/default.nix]: Sorry, I should have been more specific:
* the date should be the date of the commit you package, so as to be meaningful
* the date should include the leading 0 of months and days, so as to be increasing across updates
here: 20190211

Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: Now it should be ok

File: pkgs/servers/demoit/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  src = fetchFromGitHub {
+    owner = "dgageot";
+    repo = "demoit";
+    rev = "391ed33bfb6b459689e0f879dcf071bb4eee1608";
+    sha256 = "13jk1r4xillcwf4biqr0cyvdvm62kg808pq8fxlqsd47nw3l8pp1";
+  };
+
+  meta = with stdenv.lib; {
+    description = "Live coding demos without Context Switching";
+    homepage = "https://github.com/dgageot/demoit";
```
Comment (aanderse) [inline pkgs/servers/demoit/default.nix]: No need to quote this.

**Review by aanderse — CHANGES_REQUESTED** [review]

File: pkgs/servers/demoit/default.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+buildGoPackage rec {
+  name = "demoit-${version}";
+  version = "20190215";
+  goPackagePath = "github.com/dgageot/demoit";
+
+  src = fetchFromGitHub {
+    owner = "dgageot";
+    repo = "demoit";
+    rev = "391ed33bfb6b459689e0f879dcf071bb4eee1608";
+    sha256 = "13jk1r4xillcwf4biqr0cyvdvm62kg808pq8fxlqsd47nw3l8pp1";
```
Comment (aanderse) [inline pkgs/servers/demoit/default.nix]: Tried to build, saw this:
```
hash mismatch in fixed-output derivation '/nix/store/swzqsc7yrwair0dsvfx6nkf5dfkdgp45-source':
  wanted: sha256:13jk1r4xillcwf4biqr0cyvdvm62kg808pq8fxlqsd47nw3l8pp1
  got:    sha256:03h0ad3v9xfigrppbcxiykzgcrv3dnarm7m1vixi2j4n1340f7q6
```

File: pkgs/servers/demoit/default.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{ stdenv
+, buildGoPackage
+, fetchFromGitHub
+}:
+
+buildGoPackage rec {
+  name = "demoit-${version}";
```
Comment (aanderse) [inline pkgs/servers/demoit/default.nix]: Please change to `pname = "demoit";`

File: pkgs/servers/demoit/default.nix [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  version = "20190215";
```
Comment (aanderse) [inline pkgs/servers/demoit/default.nix]: https://nixos.org/nixpkgs/manual/
>  If a package is not a release but a commit from a repository, then the version part of the name must be the date of that (fetched) commit. The date must be in "YYYY-MM-DD" format. Also append "unstable" to the name - e.g., "pkgname-unstable-2014-09-23"

File: pkgs/servers/demoit/default.nix [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  goPackagePath = "github.com/dgageot/demoit";
+
+  src = fetchFromGitHub {
+    owner = "dgageot";
+    repo = "demoit";
+    rev = "391ed33bfb6b459689e0f879dcf071bb4eee1608";
+    sha256 = "13jk1r4xillcwf4biqr0cyvdvm62kg808pq8fxlqsd47nw3l8pp1";
+  };
+
+  meta = with stdenv.lib; {
```
Comment (aanderse) [inline pkgs/servers/demoit/default.nix]: Please add a maintainer.

**Review by aanderse — CHANGES_REQUESTED** [review]
Please rewrite git history as well.

File: pkgs/servers/demoit/default.nix (hunk 6 above)
Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: added

File: pkgs/servers/demoit/default.nix (hunk 5 above)
Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: set version to "unstable-YYYY-DD-MM"

File: pkgs/servers/demoit/default.nix (hunk 4 above)
Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: done

File: pkgs/servers/demoit/default.nix (hunk 3 above)
Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: strange behavior, I updated to latest commit

File: pkgs/servers/demoit/default.nix (hunk 2 above)
Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: fixed

**Comment by aanderse** [comment]
@freezeboy please rewrite the git history into 2 commits: the first to add yourself as a maintainer, and the second to be adding the package.

File: pkgs/servers/demoit/default.nix (hunk 3 above)
Comment (aanderse) [reply pkgs/servers/demoit/default.nix]: @freezeboy how are you getting your sha256 hashes? For the new revision you listed you should run this: `nix-prefetch-git --url https://github.com/dgageot/demoit.git --rev ec70fbdf5a5e92fa1c06d8f039f7d388e0237ba2` which gives `01584cxlnrc928sw7ldmi0sm7gixmwnawy3c5hd79rqkw8r0gbk0`.

Comment (freezeboy) [reply pkgs/servers/demoit/default.nix]: Well I used the suggestion with the conflict ... anyway, I forced-pushed a fix and I checked it builds.

**Comment by aanderse** [comment]
@GrahamcOfBorg build demoit

**Pull request merged**